[net9.0] Update net9.0 with main

rmarinho opened the pull request:
### Description of Change

Port latest changes from main to net9, makes tests green!

hartez committed:
Clean up async assertions

hartez committed:
Replace more Wait() instances

hartez committed:
Fix incorrect call in memory tests

hartez committed:
Hopefully found the last remaining Wait references

hartez committed:
Fix searchbar cancel button color test on Windows

hartez committed:
Ensure views have size in Android tests so we know they've been laid out

hartez committed:
Ensure image in stream test actually loads before checking color

hartez committed:
Simplify handler GC tests

PureWeen committed:
- fix DetailsViewPopOverLayoutIsCorrectForIdiom

Foda committed:
[Windows] Fix verification of test count in CI (#19697)

* Test ignore

* Test off-by-one

* Test print out missing category

* Cleanup

---------

Co-authored-by: Mike Corsaro <mikecorsaro@microsoft.com>

eltociear committed:
Update eng/README.md (#19735)

avalable -> available

rmarinho committed:
Clean up async assertions (#19385)

### Description of Change

Replace Wait() calls with method which Asserts the results of the
Task<bool> so we have a clearer idea where the tests fail; remove extra
post-Wait() assertions. Move non-extension methods to AssertHelpers.
Remove random timings.

### Issues Fixed

Fixes #19376

emaf committed:
[iOS] Fix Flyout layout (#19125)

* [iOS] Fix Flyout Header breaking clicking over items

The Flyout Header height was being defined as `ArrangedHeaderViewHeightWithNoMargin` + `HeaderTopMargin`, but the `HeaderView.MeasuredHeight` (value on which `ArrangedHeaderViewHeightWithNoMargin` is based on) does include the margin so we were accounting for it twice and making the header go over the items below it, making them not clickable. When there is no margin, `HeaderTopMargin` represents the safe area and it should not be accounted for either.

Fixes https://github.com/dotnet/maui/issues/17965

* Rename confusing properties

Rename `ArrangedHeaderViewHeightWithNoMargin` and `MeasuredHeaderViewHeightWithNoMargin`, to `ArrangedHeaderViewHeightWithMargin` and `MeasuredHeaderViewHeightWithMargin`, since both include the top margin value: https://github.com/dotnet/maui/blob/main/src/Core/src/Layouts/LayoutExtensions.cs#L27.

* Fix tests

* Fix safe area margin bottom value

* Fix margins

* Polish flyout header and content offsets

* Remove extra space

* Simplify content layout

* Small code cleanup

* Fix condition

* Simplify test

* More tests

* Test fix

* Add ArrangedHeaderViewHeightWithOutMargin

* Broken

* Fix scrollview inset

* Fix all tests

* Fix scroll view content inset and expand tests

* Fix new test cases for Android

* Skip FlyoutHeaderContentAndFooterAllMeasureCorrectly for Android since there is a bug that makes tests fail

* Enable more tests for iOS

* Honor IgnoreSafeArea

* Add missing null check

* Fix CollectionView support

* Add check by ItemsView for CV

* Add tests scrolling using different content control types

* Log more info on new tests

* Use requested value for assertion

* Consider epsilon for scrolled position

* Better test logging

* Fix spacing

* Another indenting issue

* Fix namespaces

MartyIX committed:
[macOS] Fix `GesturePlatformManager.CalculatePosition` when asked for a relative position (#19371)

* Fix

* Test

* Apply feedback

* Switch Button to BoxView

* - fix tests

* - add upper limit constraints

* - switch to wait for element

* Update src/Controls/samples/Controls.Sample.UITests/Issues/Issue19329.xaml

Co-authored-by: MartyIX <203266+MartyIX@users.noreply.github.com>

* - switch to click for catalyst

* Update src/Controls/tests/UITests/Tests/Issues/Issue19329.cs

Co-authored-by: MartyIX <203266+MartyIX@users.noreply.github.com>

* - add comments

---------

Co-authored-by: Shane Neuville <shneuvil@microsoft.com>
Co-authored-by: Shane Neuville <shane94@hotmail.com>

hartez committed:
Fix disappearing cell issue for TableView on Android (#19723)

Disables the native cell recycling mechanisms in favor of the renderer/handler reuse that's already built into
.NET MAUI. Adds special casing to ensure that the behavior changes only apply to TableView; similar problems may
exist in ListView, but that's beyond the scope of this change.

These changes also break down the TableView's cell creation methods for easier maintenance.

dotnet-maestro[bot] committed:
[main] Update dependencies from dotnet/xharness (#19585)

* Update dependencies from https://github.com/dotnet/xharness build 20231220.1

Microsoft.DotNet.XHarness.CLI , Microsoft.DotNet.XHarness.TestRunners.Common , Microsoft.DotNet.XHarness.TestRunners.Xunit
 From Version 9.0.0-prerelease.23616.1 -> To Version 9.0.0-prerelease.23620.1

* Update dependencies from https://github.com/dotnet/xharness build 20231220.1

Microsoft.DotNet.XHarness.CLI , Microsoft.DotNet.XHarness.TestRunners.Common , Microsoft.DotNet.XHarness.TestRunners.Xunit
 From Version 9.0.0-prerelease.23616.1 -> To Version 9.0.0-prerelease.23620.1

* Try update catalyst to use the xharness installed

---------

Co-authored-by: dotnet-maestro[bot] <dotnet-maestro[bot]@users.noreply.github.com>
Co-authored-by: Rui Marinho <me@ruimarinho.net>

PureWeen committed:
Update bug-report.yml with 8.0.5 (#19795)

PureWeen committed:
Move GroupListViewHeaderIndexOutOfRange to Appium (#19799)

hartez committed:
Enable string search and comparison CA rules in Core (#19796)

github-actions[bot] committed:
[create-pull-request] automated change (#19769)

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

StephaneDelcroix committed:
[X] passing test for 18976 (#19785)

- fixes #18976

hartez committed:
Handle setting of DesiredSize in Measure implementations, rather than MeasureOverride (#19794)

hartez committed:
Enable CA1822 in Core and convert eligible methods to static (#19800)

* Enable CA1822 in Core and convert eligible methods to static

* Tizen fixes

* Disable CA1822 for partial method

github-actions[bot] committed:
[create-pull-request] automated change (#19813)

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

Commit:
Fix issue 18110

PureWeen committed:
Fix SetProgressViewOffset on SwipeRefreshLayout (#19767)

* Fix SetProgressViewOffset on SwipeRefreshLayout

* - update refresh icon to a better color to trigger screen shot comparison

* - remove extra code

* - add better exception

hartez committed:
Fix incorrect defaults when adding non-BindableObject as AbsoluteLayout child (#19839)

rmarinho committed:
Merge branch 'main' into net9-main

# Conflicts:
#	.config/dotnet-tools.json
#	eng/Version.Details.xml
#	eng/Versions.props
#	src/Controls/src/Core/Compatibility/Handlers/Shell/iOS/ShellFlyoutHeaderContainer.cs
#	src/Controls/src/Core/Compatibility/Handlers/Shell/iOS/ShellFlyoutLayoutManager.cs
#	src/Controls/src/Core/Compatibility/Handlers/TableView/Android/TableViewModelRenderer.cs
#	src/Controls/tests/DeviceTests/Elements/Shell/ShellFlyoutHeaderScrollTestCases.cs
#	src/Controls/tests/DeviceTests/Elements/Shell/ShellFlyoutTests.cs
#	src/Controls/tests/DeviceTests/Elements/Shell/ShellFlyoutTests.iOS.cs
#	src/Controls/tests/DeviceTests/Elements/Shell/ShellTests.Android.cs
#	src/Controls/tests/DeviceTests/Elements/Shell/ShellTests.iOS.cs
#	src/Controls/tests/UITests/Tests/Issues/Issue19329.cs
#	src/Core/tests/DeviceTests/Memory/MemoryTestFixture.cs
#	src/TestUtils/src/DeviceTests/AssertHelpers.cs
#	src/TestUtils/src/DeviceTests/AssertionExtensions.Android.cs
#	src/TestUtils/src/DeviceTests/AssertionExtensions.cs

drasticactions committed:
[Windows] Show the Raw HTML for Html Label in case of error (#17470)

* Show the Raw HTML in case of error

* Try to add UITest for Broken HTML

jsuarezruiz commented on src/Controls/tests/DeviceTests/Elements/Shell/ShellTests.iOS.cs:
```diff
@@ -16,9 +16,9 @@
 using Microsoft.Maui.Platform;
 using UIKit;
 using Xunit;
+using static Microsoft.Maui.DeviceTests.AssertHelpers;
```
```
D:\a\_work\1\s\src\Controls\tests\DeviceTests\Elements\Shell\ShellTests.Windows.cs(20,14): error CS0105: The using directive for 'AssertHelpers' appeared previously in this namespace [D:\a\_work\1\s\src\Controls\tests\DeviceTests\Controls.DeviceTests.csproj::TargetFramework=net9.0-windows10.0.20348]
D:\a\_work\1\s\src\Controls\tests\DeviceTests\Elements\Window\WindowTests.Windows.cs(14,14): error CS0105: The using directive for 'AssertHelpers' appeared previously in this namespace [D:\a\_work\1\s\src\Controls\tests\DeviceTests\Controls.DeviceTests.csproj::TargetFramework=net9.0-windows10.0.20348]
D:\a\_work\1\s\src\Core\tests\DeviceTests\Handlers\Button\ButtonHandlerTests.Android.cs(10,14): error CS0105: The using directive for 'AssertHelpers' appeared previously in this namespace [D:\a\_work\1\s\src\Core\tests\DeviceTests\Core.DeviceTests.csproj::TargetFramework=net9.0-android]
D:\a\_work\1\s\src\Controls\tests\DeviceTests\Elements\CollectionView\CollectionViewTests.Windows.cs(14,14): error CS0105: The using directive for 'AssertHelpers' appeared previously in this namespace [D:\a\_work\1\s\src\Controls\tests\DeviceTests\Controls.DeviceTests.csproj::TargetFramework=net9.0-windows10.0.19041]
D:\a\_work\1\s\src\Controls\tests\DeviceTests\Elements\Shell\ShellTests.Windows.cs(20,14): error CS0105: The using directive for 'AssertHelpers' appeared previously in this namespace [D:\a\_work\1\s\src\Controls\tests\DeviceTests\Controls.DeviceTests.csproj::TargetFramework=net9.0-windows10.0.19041]
D:\a\_work\1\s\src\Controls\tests\DeviceTests\Elements\Window\WindowTests.Windows.cs(14,14): error CS0105: The using directive for 'AssertHelpers' appeared previously in this namespace [D:\a\_work\1\s\src\Controls\tests\DeviceTests\Controls.DeviceTests.csproj::TargetFramework=net9.0-windows10.0.19041]
    5775 Warning(s)
    7 Error(s)

```

mikeparker104 committed:
[Catalyst] Allow setting of CancelButtonColor on SearchBar when using Mac Idiom (#19838)

### Description of Change

Update to ```SearchBarExtensions``` so ```UpdateCancelButton``` sets the
```TintColor``` on the Cancel ```UIButton``` when using Mac idiom.
Device test added verifying the changes for Mac Catalyst and iOS.

### Issues Fixed

Fixes [#18110](https://github.com/dotnet/maui/issues/18110)

tarasverq committed:
added optional ReceiveSms permission (#19659)

tj-devel709 committed:
iOS Keyboard Scrolling Improvements and UITests (not from fork) (#17670)

* Squash all the iOS Keyboard Improvements

* delay was enough manually, but not for tests

* remove debugging

* Add logic for Disconnect to be more useful

* fix spacing from github merge conflict commit

* accidently removed a var declaration

* Improve UITests and add double scroll for LargeTitles

* Add UITest for the grid star

* Style changes

* more style

* add fixtureTeardown

* only check for LargeTitles in portrait mode

* Add changes for LargeTitles on iPad

* remove the popover code until we get tests

rmarinho committed:
[ci] Bump XCode to 15.2.0 (#19836)

* [ci] Bump XCode to 15.1.0

* Go to Xcode 15.2

* New xcode 15.1 doesn't support 14.5 simulators

hartez committed:
Add missing AbsoluteLayout section to layout architecture doc (#19841)

rmarinho committed:
[ci] Don't install xharness globally (#19863)

* dont install xharness globally

* Fix restore

jamesmontemagno committed:
Update README.md (#19797)

jonathanpeppers committed:
[xamlg] improve error for Xamarin.Forms namespace (#19683)

* [xamlg] improve error for Xamarin.Forms namespace

Fixes: https://github.com/dotnet/maui/issues/18637

Using the old namespace `http://xamarin.com/schemas/2014/forms` was
causing a NRE:

    System.NullReferenceException: Object reference not set to an instance of an object.
    at Microsoft.Maui.Controls.Xaml.XmlTypeXamlExtensions.GetTypeReference[T](XmlType xmlType, IEnumerable`1 xmlnsDefinitions, String defaultAssemblyName, Func`2 refFromTypeInfo)
    at Microsoft.Maui.Controls.SourceGen.CodeBehindGenerator.GetTypeNameFromCustomNamespace(XmlType xmlType, Compilation compilation, AssemblyCaches caches)
    at Microsoft.Maui.Controls.SourceGen.CodeBehindGenerator.GetTypeName(XmlType xmlType, Compilation compilation, AssemblyCaches caches)
    at Microsoft.Maui.Controls.SourceGen.CodeBehindGenerator.TryParseXaml(SourceText text, String uid, Compilation compilation, AssemblyCaches caches, CancellationToken cancellationToken, String targetFramework, String& accessModifier, String& rootType, String& rootClrNamespace, Boolean& generateDefaultCtor, Boolean& addXamlCompilationAttribute, Boolean& hideFromIntellisense, Boolean& xamlResourceIdOnly, String& baseType, IEnumerable`1& namedFields, Exception& exception)
    at Microsoft.Maui.Controls.SourceGen.CodeBehindGenerator.GenerateXamlCodeBehind(ProjectItem projItem, Compilation compilation, SourceProductionContext context, AssemblyCaches caches)
    at Microsoft.Maui.Controls.SourceGen.CodeBehindGenerator.<>c.<Initialize>b__1_1(SourceProductionContext sourceProductionContext, ValueTuple`3 provider)
    at Microsoft.CodeAnalysis.UserFunctionExtensions.<>c__DisplayClass3_0`2.<WrapUserAction>b__0(TInput1 input1, TInput2 input2, CancellationToken token)

In f939049f, I inadvertently introduced this NRE, but even after fixing
it, I would instead get:

    FormsNamespace.xaml.sg.cs(15,48): error CS1001: Identifier expected

Due to the namespace, type, and base type all being `null` in this case,
the generated C# is invalid.

Instead of generating C# *at all*, let's emit an error message that says
to use the correct namespace.

I also provided a `Location` for the error message, so we will now know
what file the error originated from.

I attempted to write some test here, but adding an invalid file like
this causes the XAML unit tests project to fail to build. It feels like
we actually need to create a unit test project for the source generator
-- as there isn't one now.

* Add `[Obsolete]` to `FormsUri`

rmarinho committed:
Merge branch 'main' into net9-main

# Conflicts:
#	eng/pipelines/common/maui-templates.yml
